Add `ComponentSelector` feature: IS portion

GabrielKS opened the pull request:
~This draft PR will get filled in as @tengis-nrl and I move the ComponentSelector features from PowerAnalytics to InfrastructureSystems.~ Ready for review! Adds an immutable, lazy, system-independent representation of a grouped set of components. For more details and a demonstration, see https://github.nrel.gov/gkonars/PowerAnalytics-demos/blob/main/component_selector_pr_demo.ipynb.

tengis-nrl commented:
<img width="451" alt="image" src="https://github.com/NREL-Sienna/InfrastructureSystems.jl/assets/156373061/65d7e5bb-ab46-4e3a-8443-72ebf53b9585">

jd-lara commented on jl_5D19.tmp:
Why are all these tmp files checked in?

tengis-nrl replied:
Sorry, I forgot to remove those. Let me do it now.

GabrielKS replied:
There remains a temp file

tengis-nrl replied:
@GabrielKS can you find that file and delete it please? I can't find the folder or the file itself on VS Code.

GabrielKS replied:
@tengis-nrl done

GabrielKS commented:
We need working implementations of `get_components` and `get_subselectors` here, not just in PowerSystems, because the ComponentSelector feature is meant to be usable without PowerSystems. You'll note a TODO [here](https://github.com/NREL-Sienna/InfrastructureSystems.jl/commit/8ccdb70a3a5a0b71321817c992b7d0a111829f92#diff-781d82a2ab621d91ba59dbb664bc65d38f895d1541a91d2aae94d9ae0694ba48R3) about making `get_available` optional; if you don't think we have time to do that now, I think the IS version of `get_components` should just not check availability and the PSY version should.

GabrielKS commented:
Rebased onto `main` and tests pass.

GabrielKS commented:
This is now ready for in-depth code review! For a demonstration, see https://github.nrel.gov/gkonars/PowerAnalytics-demos/blob/main/component_selector_pr_demo.ipynb.

daniel-thom reviewed (commented):
This is a first pass to get some pedantic stuff out of the way.

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,223 @@
+# SingleComponentSelector
+"ComponentSelector that wraps a single Component."
+struct SingleComponentSelector <: ComponentSelectorElement
+    component_subtype::Type{<:InfrastructureSystemsComponent}
```
Agreed, replaced lots of occurrences of `subtype` with `type`

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,223 @@
+collection with the given Component's subtype and name.
+"""
+select_components(
+    component_ref::InfrastructureSystemsComponent,
```
Done

GabrielKS commented on src/common.jl:
```diff
@@ -41,6 +41,9 @@ const CONSTANT = Float64
 const LIM_TOL = 1e-6
 const XY_COORDS = @NamedTuple{x::Float64, y::Float64}
 
+"Delimeter to use when constructing qualified names like component_type__component_name."
+const COMPONENT_NAME_DELIMETER = "__"
```
Done

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+`DynamicallyGroupedComponentSelector`
+"""
+function get_groups(
+    e::DynamicallyGroupedComponentSelector,
```
Agreed, done (`e` was a vestige of these being called "entities")

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+    name::Union{String, Nothing} = nothing,
+) = NameComponentSelector(component_subtype, component_name, name)
+"""
+Make a `ComponentSelector` from a component reference, pointing to components in any
```
Done

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+default_name(e::SubtypeComponentSelector) = subtype_to_string(e.component_subtype)
+
+# Contents
+function get_components(e::SubtypeComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
```
So there's this analogy:

- Eager: call `get_components(type, system)`. Lazy: call `make_selector(type)` and later call `get_components` on it.
- Eager: call `get_component(type, system, name)`. Lazy: call `make_selector(type, name)` and later call `get_component` on it.
- Eager: call `get_components(filter_func, type, system)`. Lazy: call `make_selector(type, filter_fn)` and later call `get_components` on it (granted this means I should probably reorder the parameters to that `make_selector` method).

so in that sense, my `filter_fn` below is analogous to the existing `filter_func` and this `filterby` kwarg is something else, just an extra feature to satisfy a special case (I talk about this in the demo notebook).

I'll rename `filter_fn` to `filter_func`. My rationale for giving `filter_fn` and `filterby` different names was that they *are* different things, but maybe that's not a good argument, thoughts?

I don't really foresee users wanting to use a `do` block with this, and I think the fact that `filterby` is an optional add-on to the existing selector interface justifies its inclusion as a kwarg rather than the first parameter, but I don't think it hurts to do both; I can add a wrapper method where the first parameter is `filterby`.

GabrielKS replied:
Update, I think every part of my interface that takes a `filterby` or `filter_fn` (now `filter_func`) now has the option of putting that first.

GabrielKS replied:
I've renamed `filterby` to `scope_limiter` to reduce confusion with `filter_func`. The mental model is maybe like a [universe](https://en.wikipedia.org/wiki/Universe_(mathematics)) in math: it can sometimes make sense to limit the scope of objects you are considering from literally everything to some collection that makes sense in the context (e.g., only `available` components) before doing the actually interesting filtering. Open to alternate naming suggestions, also considered `domain_limiter`.

GabrielKS replied:
Resolving this as `scope_limiter` is no longer part of the user interface.

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+Canonical way to turn an `InfrastructureSystemsComponent` specification/instance into a
+unique-per-system string.
+"""
+component_to_qualified_string(
```
Looks like the difference is `make_label` yields `ThermalStandard: 322_CT_6` whereas `component_to_qualified_string` yields `ThermalStandard__322_CT_6`. Not to wade into the Symbol debate again, but I want my dataframe columns to be dot-accessible and such…

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+# because only PSY.get_components, not IS.get_components, is defined for PSY.System, we need
+# to be able to easily override these methods. See
+# https://github.com/NREL-Sienna/InfrastructureSystems.jl/issues/388
+const SYSTEM_LIKE = Union{Components, SystemData}
```
You're not worried that people will think they *are* real types and go "what's a `SystemLike`?"?

daniel-thom replied:
I would argue that `SystemLike` _is_ effectively a type.

GabrielKS replied:
Fair enough; done.

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+implementation of `get_groups`.
+=#
+"`PluralComponentSelector`s whose grouping is determined by a `groupby` field."
+abstract type DynamicallyGroupedComponentSelector <: PluralComponentSelector end
```
Dynamic roughly means that grouping is on the fly: when you call `get_groups`, it will take stock of what components actually exist in the system and make the groups as necessary. `ListComponentSelector` has groups, but they're not dynamic: if you construct a `ListComponentSelector` with two groups, it will always have those two groups no matter the system.

(in fact, it would be nice to be able to call `get_groups` on non-dynamically-grouped selectors *without* having to pass in the system, but that's for the next version to worry about)

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+
+# SubtypeComponentSelector
+"`PluralComponentSelector` represented by a type of component."
+struct SubtypeComponentSelector <: DynamicallyGroupedComponentSelector
```
Is `TypeComponentSelector` okay, to go with my naming convention?

GabrielKS replied:
(renamed to `TypeComponentSelector` for now)

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+# CONCRETE SUBTYPE IMPLEMENTATIONS
+# NameComponentSelector
+"`ComponentSelector` that refers by name to at most a single component."
+struct NameComponentSelector <: SingularComponentSelector
```
It is exposed to users through `make_selector`. I could imagine an analysis where you have results from a few slightly different versions of the system and you want to analyze the behavior of one particular component across the scenarios.

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+Get the name of the `ComponentSelector`. This is either the default name or a custom name
+passed in at creation time.
+"""
+get_name(e::ComponentSelector) = (e.name !== nothing) ? e.name : default_name(e)
```
It would slightly simplify the code (though would require manual constructors…). I remember thinking about this a long time ago, I'll try to reconstruct why I did it this way…

GabrielKS replied:
Okay, I bit the bullet and did manual constructors. `default_name` is eliminated now

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+    component_to_qualified_string(e.component_subtype, e.component_name)
+
+# Contents
+function get_components(e::NameComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
```
(see below)

GabrielKS replied:
(done)

daniel-thom commented on src/component_selector.jl:
```diff
@@ -0,0 +1,371 @@
+
+    return [
+        make_selector(
+            make_selector.(components[partition_results .== Ref(group_result)])...;
```
Consider whether this is too clever for the average user. It will be read many times.

GabrielKS replied:
Refactored with a helper function and no `Ref`.

daniel-thom commented on src/component_selector.jl:
```diff
@@ -0,0 +1,295 @@
+function get_components(e::FilterComponentSelector, sys::SYSTEM_LIKE; filterby = nothing)
+    components = get_components(e.filter_fn, e.component_subtype, sys)
+    isnothing(filterby) && (return components)
+    return Iterators.filter(filterby, components)
```
This makes the function type unstable. We created `FlattenIteratorWrapper` so that we know the length. We shouldn't return different types based on a kwarg.

GabrielKS replied:
1. Agreed, I'll make these type stable.
2. Looks like not really, not after compilation anyway:
<img width="929" alt="Screenshot 2024-09-30 at 3 20 06 PM" src="https://github.com/user-attachments/assets/447a0b20-45bb-436a-98ce-056d53b78e93">
(this is on the massive EI system)

GabrielKS replied:
`get_components` now always returns `FlattenIteratorWrapper`. I also ended up combining the two filter functions into one as an implementation detail (not for performance, it just ended up being more convenient with `FlattenIteratorWrapper`).

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,371 @@
+    sys;
+    filterby = nothing,
+)
+    validate_groupby(selector.groupby)
```
Done now

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,408 @@
+ - `scope_limiter`: optional filter function to limit the scope of components under consideration
+"""
+get_components(
+    scope_limiter::Union{Nothing, Function},
```
It is mostly a stylistic thing, I could concoct a scenario where the user is broadcasting/looping and splatting where it would make sense (when you get to the PowerAnalytics PR you'll see some cases where this does actually happen) but that does seem unlikely in this case

GabrielKS replied:
Resolving as obsolete: `scope_limiter` is no longer part of the user interface.

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,408 @@
+ - `scope_limiter`: optional filter function to limit the scope of components under consideration
+"""
+get_groups(
+    scope_limiter::Union{Nothing, Function},
```
See above

GabrielKS commented on src/component_selector.jl:
```diff
@@ -0,0 +1,408 @@
+"""
+make_selector(
+    component_type::Type{<:InfrastructureSystemsComponent},
+    filter_func::Function;
```
There is a signature (line 386) where it is first. Is this compromise where I have both signatures acceptable? It just seems really inconsistent to me to have the user specify the component type first in every case except the filter function case. This way the `do` block is supported but I also get my consistency.

daniel-thom replied:
My concern is consistency in our API. Currently, there is only one way to call `get_components` with a filter function. That function has to be first. This PR is introducing a second pattern where you can call `get_components` and specify the filter function or scope limiter as a keyword argument. If people see this pattern being used they may try to use it with the existing `get_components` and then complain that it doesn't work. I don't want to have to then go back and change that method to support this new pattern. My suggestion is that we continue to provide a single way to specify filter functions - and that it follow the Julia convention.

GabrielKS replied:
For the record, @daniel-thom, @jd-lara, @sourabhdalvi, @akrivi and I discussed this synchronously and came to the following resolution:

1. Eliminate the `scope_limiter` feature. Instead:
   - When the user calls `get_components(selector, system)`, return all components that satisfy the selector (no availability filtering)
   - When the user calls `get_components(selector, results)`, return only the available components
   - When the user calls `get_available_components(selector, system)`, return only the available components
2. Follow the Julia convention of always putting the function first; remove the methods where the function is not first (I was outvoted)

GabrielKS replied:
From the user's point of view, this is done. There is a hidden `scope_limiter` kwarg to aid in implementation and extension.

GabrielKS commented on src/component_container.jl:
```diff
@@ -0,0 +1,71 @@
+abstract type ComponentContainer <: InfrastructureSystemsContainer end
+
+# For each of these functions, `ComponentContainer` concrete subtypes MUST either implement
+# appropriate methods or accept the default if it exists.
```
@daniel-thom these are the only ones I need for `ComponentSelector` and I'm trying to avoid further scope creep. But maybe we also want `iterate_components`, etc. here?

daniel-thom replied:
Mabye. I'm definitely OK leaving it out.

GabrielKS commented on src/system_data.jl:
```diff
@@ -17,7 +17,7 @@ const SERIALIZATION_METADATA_KEY = "__serialization_metadata__"
 Container for system components and time series data
 """
-mutable struct SystemData <: InfrastructureSystemsType
+mutable struct SystemData <: ComponentContainer
```
`ComponentContainer <: InfrastructureSystemsType` so this is backwards compatible

GabrielKS commented:
@daniel-thom Ready for another review. The main changes since last review are:

1. Cleanup of the `scope_limiter` situation. More details above, but the gist is that this is no longer part of the user interface, its functionality is served by `get_components` vs. `get_available_components`.
2. Addition of `rebuild_selector` and the supporting type `RegroupedComponentSelector` to fulfill a request from @akrivi
3. Addition of `ComponentContainer` with a standard interface `get_components`-type stuff

daniel-thom commented on src/component_selector.jl:
```diff
@@ -0,0 +1,434 @@
+function get_components(selector::FilterComponentSelector, sys; kwargs...)
+    # Short-circuit-evaluate the `scope_limiter` first so `filter_func` may refer to
+    # component attributes that do not exist in components outside the scope
+    scope_limiter = get(kwargs, :scope_limiter, nothing)
```
It seems like you want `scope_limiter` to be part of the public interface, or at least you may be setting it in PowerAnalytics. Is that correct?

daniel-thom commented on src/component_selector.jl:
```diff
@@ -0,0 +1,434 @@
+Get the available components of the collection that make up the `ComponentSelector`.
+"""
+get_available_components(selector::ComponentSelector, sys) =
+    get_components(selector, sys; scope_limiter = x -> get_available(sys, x))
```
I'm a bit concerned that people will use this as example code. And then eventually ask why scope_limiter isn't a kwarg in the System get_components method. Wouldn't it be better to follow the Julia convention and pass the "scope_limiter" as the first argument?

GabrielKS replied:
The goals with this design are to (a) keep `scope_limiter` out of the public interface, as we resolved earlier, while (b) keeping it in a sort of "developer interface" so that it can be used for `get_available_components`. Absent another layer of indirection (have each `ComponentSelector` subtype implement `_get_components` with a `scope_limiter` and have `get_components` and `get_available_components` call that), putting `scope_limiter` as the first argument would more or less _force_ it be part of the public interface: it would explicitly appear in method lists and such. Do you think it's worth it to add that layer of indirection, or is the current solution (perhaps with a comment added here that users are not to take this source code as example code?) good enough?

daniel-thom reviewed (approved):
LGTM
